lib/cn/cn_tree.c

```diff
@@ -1356,71 +1359,103 @@ cn_comp_update_kvcompact(struct cn_compaction_work *work, struct kvset *new_kvse
+cn_input_kvsets_delete(struct cn_compaction_work *work)
 {
-    struct cn_tree *         tree = work->cw_tree;
-    struct cn_tree_node *    pnode = work->cw_node;
-    struct kvset_list_entry *le, *tmp;
     struct list_head         retired_kvsets;
-    struct cn_tree_node *    node;
+    struct cn_tree          *tree = work->cw_tree;
+    struct cn_tree_node     *pnode = work->cw_node;
+    struct kvset_list_entry *le, *tmp;
+    struct cndb_txn         *tx;
 
-    if (ev(work->cw_err))
-        return;
+    assert(!work->cw_err);
 
     INIT_LIST_HEAD(&retired_kvsets);
 
-    rmlock_wlock(&tree->ct_lock);
-    {
-        for (uint i = 0; i < work->cw_outc; i++) {
-            if (kvsets[i]) {
-                node = work->cw_output_nodev[i];
-                assert(node);
-                kvset_list_add(kvsets[i], &node->tn_kvset_list);
-                node->tn_cgen++;
-            }
-        }
-
-        /* Advance the change generation on the spill source node
-         * to ensure it is reevaluated by csched/sp3_dirty_node().
-         */
-        pnode->tn_cgen++;
+    work->cw_err = cndb_record_txstart(work->cw_tree->cndb, 0, CNDB_INVAL_INGESTID, CNDB_INVAL_HORIZON,
+                                       0, work->cw_kvset_cnt, &tx);
+    if (work->cw_err)
+        goto done;
 
-        /* Move old kvsets from parent node to retired list.
-         * Asserts:
-         * - Each input kvset just spilled must still be on pnode's kvset list.
-         * - The dgen of the oldest input kvset must match work struct dgen_lo
-         *   (i.e., concurrent spills from a node must be committed in order).
-         */
-        for (uint i = 0; i < work->cw_kvset_cnt; i++) {
-            assert(!list_empty(&pnode->tn_kvset_list));
-            le = list_last_entry(&pnode->tn_kvset_list, struct kvset_list_entry, le_link);
-            assert(i > 0 || work->cw_dgen_lo == kvset_get_dgen(le->le_kvset));
-            list_del(&le->le_link);
-            list_add(&le->le_link, &retired_kvsets);
-        }
+    le = work->cw_mark;
+    for (uint i = 0; i < work->cw_kvset_cnt; i++) {
+        assert(le);
+        assert(tx);
 
-        cn_tree_samp(tree, &work->cw_samp_pre);
+        work->cw_err = kvset_delete_log_record(le->le_kvset, tx);
+        if (work->cw_err)
+            goto done;
 
-        cn_tree_samp_update_spill(tree, pnode);
+        le = list_prev_entry(le, le_link);
+    }
 
-        cn_tree_samp(tree, &work->cw_samp_post);
+    rmlock_wlock(&tree->ct_lock);
 
-        assert(!work->cw_have_token);
+    /* Move old kvsets from parent node to retired list.
+     * Asserts:
+     * - Each input kvset just spilled must still be on pnode's kvset list.
+     * - The dgen of the oldest input kvset must match work struct dgen_lo
+     *   (i.e., concurrent spills from a node must be committed in order).
+     */
 
-        atomic_sub_rel(&pnode->tn_busycnt, (1u << 16) + work->cw_kvset_cnt);
+    for (uint i = 0; i < work->cw_kvset_cnt; i++) {
```

Comment:
I think we can replace the lines 1433 - 1439 with the below?

le = list_next_entry(le, le_link);
list_trim(&retired_kvsets, &pnode->tn_kvset_list, &le->le_link);